crates/next-core/src/middleware.rs

```diff
@@ -34,11 +35,66 @@ pub async fn get_middleware_module(
     let is_proxy = userland_path.file_stem() == Some("proxy");
     let page_path = if is_proxy { "/proxy" } else { "/middleware" };
 
+    // Validate that the module has the required exports
+    if let Some(ecma_module) =
+        Vc::try_resolve_sidecast::<Box<dyn EcmascriptChunkPlaceable>>(*userland_module).await?
+    {
+        let exports = ecma_module.get_exports().await?;
+
+        // Check if the module has the required exports
+        let has_valid_export = match &*exports {
+            // ESM modules - check for named or default export
+            EcmascriptExports::EsmExports(esm_exports) => {
+                let esm_exports = esm_exports.await?;
+                let has_default = esm_exports.exports.contains_key(&rcstr!("default"));
+                let expected_named = if is_proxy {
+                    rcstr!("proxy")
+                } else {
+                    rcstr!("middleware")
+                };
+                let has_named = esm_exports.exports.contains_key(&expected_named);
+                has_default || has_named
+            }
+            // CommonJS modules are valid (they can have module.exports or exports.default)
+            EcmascriptExports::CommonJs | EcmascriptExports::Value => true,
+            // DynamicNamespace might be valid for certain module types
+            EcmascriptExports::DynamicNamespace => true,
+            // Empty or unknown exports are invalid
+            EcmascriptExports::EmptyCommonJs
+            | EcmascriptExports::None
+            | EcmascriptExports::Unknown => false,
+        };
+
+        if !has_valid_export {
+            let file_type = if is_proxy { "Proxy" } else { "Middleware" };
+            let function_name = if is_proxy { "proxy" } else { "middleware" };
+            // Extract just the filename for the error message
+            let file_name = userland_path
+                .path
+                .split('/')
+                .next_back()
+                .unwrap_or(&userland_path.path);
+            // Use the same error message format as the runtime check
+            bail!(
+                "The {} file \"./{}\" must export a function named `{}` or a default function.",
+                file_type,
+                file_name,
+                function_name
```

Comment:
We really shouldn't use `bail!` for issues in user code. We should instead emit an implementation of the `Issue` trait. e.g. https://github.com/vercel/next.js/blob/32476071fe331948d89a35c391eb578aed8de979/crates/next-core/src/next_shared/webpack_rules/babel.rs#L297-L348

After emitting, the logic should probably continue on as if there were no middleware file, or it should do a best-effort creation of a middleware module. You only want to `bail` if something is truly fatal and should block us from trying to collect other issues or from serving pages with the dev server.